test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs

```diff
@@ -0,0 +1,308 @@
+                            .RequireParentTag("div"))
+                    .TagMatchingRule(
+                        rule => rule
+                            .RequireTagName("strong")
+                            .RequireParentTag("div"))
+                    .Build()
+            };
+            var documentContext = TagHelperDocumentContext.Create(string.Empty, documentDescriptors);
+            var service = new DefaultTagHelperFactsService();
+
+            // Act
+            var descriptors = service.GetTagHelpersGivenTag(documentContext, "strong", "div");
+
+            // documentDescriptors
+            Assert.Equal(documentDescriptors, descriptors, TagHelperDescriptorComparer.CaseSensitive);
+        }
+
+        [Fact]
+        public void GetTagHelpersGivenParent_AllowsUnspecifiedParentTagHelpers()
+        {
+            // Arrange
+            var documentDescriptors = new[]
+            {
+                ITagHelperDescriptorBuilder.Create("TestType", "TestAssembly")
+                    .TagMatchingRule(rule => rule.RequireTagName("div"))
+                    .Build()
+            };
+            var documentContext = TagHelperDocumentContext.Create(string.Empty, documentDescriptors);
+            var service = new DefaultTagHelperFactsService();
+
+            // Act
+            var descriptors = service.GetTagHelpersGivenParent(documentContext, "p");
+
+            // Assert
+            Assert.Equal(documentDescriptors, descriptors, TagHelperDescriptorComparer.CaseSensitive);
+        }
+
+        [Fact]
+        public void GetTagHelpersGivenParent_RestrictsTagHelpersBasedOnParent()
+        {
+            // Arrange
+            var expectedDescriptors = new[]
+            {
+                ITagHelperDescriptorBuilder.Create("TestType", "TestAssembly")
+                    .TagMatchingRule(
+                        rule => rule
+                            .RequireTagName("p")
+                            .RequireParentTag("div"))
+                    .Build()
+            };
+            var documentDescriptors = new[]
+            {
+                expectedDescriptors[0],
+                ITagHelperDescriptorBuilder.Create("TestType2", "TestAssembly")
+                    .TagMatchingRule(
+                        rule => rule
+                            .RequireTagName("strong")
+                            .RequireParentTag("p"))
+                    .Build()
+            };
+            var documentContext = TagHelperDocumentContext.Create(string.Empty, documentDescriptors);
+            var service = new DefaultTagHelperFactsService();
+
+            // Act
+            var descriptors = service.GetTagHelpersGivenParent(documentContext, "div");
+
+            // Assert
+            Assert.Equal(expectedDescriptors, descriptors, TagHelperDescriptorComparer.CaseSensitive);
+        }
+
+        [Fact]
+        public void GetTagHelpersGivenParent_DoesNotReturnDuplicateDescriptors()
```

Comment:
I don't think we should do things like this. Ask a stupid question, get a stupid answer.